// gplus/index.js
var http = require("http");
var https = require("https");
var urls = { optPreLogin : 'https://accounts.google.com/login',
              optLogin : 'https://accounts.google.com/ServiceLoginAuth',
              optPostLogin : 'https://plus.google.com/',
              optPost : 'https://plus.google.com/_/'
            };

var cookies = {};
var objs = [];

function parse(url)
{
  var m = url.match(/(https?):\/\/([^\/\?]+)([^\?]+)*(\?.*)*/);
  var m2 = m[2].match(/([^:]+):?(.*)/);
  var options = {};

  options.host = m2[1];
  options.path = (m[3] ? m[3] : '/') + (m[4] ? m[4] : '');
  options.proto = m[1];
  options.port = m2[2] ? m2[2] : m[1]=='http' ? 80 : 443;
  return options;
}

function randomID(x)
{
  var id = '';

  for(var i=0;i<x;i++)
  {
    id = id + Math.floor(Math.random()*10);
  }
  return id;
}

function getCookie(options) {

  var cookie = '';
  var hostname = options.host;

  if (typeof hostname != 'string') return '';
  if (hostname.length < 1) return '';

  for (var key in cookies) {
    if (hostname.indexOf(key) > -1) {
      for (var path in cookies[key]) {
        if (options.path.indexOf(path) === 0) {
          for (var name in cookies[key][path]) {
            cookie += cookies[key][path][name]['value'] + '; ';
          }
        }
      }
    }
  }
  return cookie;
}

function cookieStore(headers,dom) {

  if (!headers['set-cookie']) return;

  headers['set-cookie'].forEach(function(cookie) {

    try {
      var domain = dom || '',
      name   = '',
      value  = '',
      expires = '',
      path   = '',
      secure = '',
      httponly = '',
      segments = cookie.split(';');

      segments.forEach(function (segment) {
        if (segment.indexOf('Path') === 0)
          path = segment.substring(5, segment.length);
        if (segment.indexOf('Domain') === 0)
          domain = segment.substring(7, segment.length);
      });
      value = segments[0];
      name = value.split('=')[0];
      cookies[domain] = cookies[domain] || {};
      cookies[domain][path] = cookies[domain][path] || {};
      cookies[domain][path][name] = cookies[domain][path][name] || {};
      cookies[domain][path][name]['value'] = value;
      cookies[domain][path][name]['expires'] = expires;
      cookies[domain][path][name]['secure'] = secure;
      cookies[domain][path][name]['httponly'] = httponly;

    } catch (e) {
      throw e;
    }
  });
}

function gPlus(){
  var _email;
  var _passwd;
  var _loginCb;
}
module.exports = gPlus;

function onPostLogin(e,r,buffer)
{

  var obj = new RegExp("AF_initDataCallback\\({[\\s\\S]*?\\);","g");
  var m = buffer.match(obj);
  for(var i=0;i<m.length;i++)
  {
    m[i] =m[i].replace("AF_initDataCallback(","").replace(');','');
    objs[i] = eval('(' + m[i] + ')');

    if(objs[i].key == "1"){
      sendid = JSON.stringify(objs[i].data[15]);
      sendid = sendid.replace(new RegExp("\"","g"),"");
    }
    if(objs[i].key == "2"){
      userid = JSON.stringify(objs[i].data[0]);
      userid = userid.replace(new RegExp("\"","g"),"");
    }
  }
  fsid = randomID(10);
  _loginCb();
}

function onLogin(e,r,b)
{

  options = parse(urls['optPostLogin']);
  options.headers = {};

  _get(options,onPostLogin);
}

function onPreLogin(e,r,b)
{

  options = parse(urls['optLogin']);
  options.form={};
  var dsh = b.match(/name=\"dsh\" id=\"dsh\" value=\"([-0-9]+)\"/);
  var galx = b.match(/name=\"GALX\"[ \t\n]+value=\"([-_a-zA-Z0-9]+)\">/);
  if(!galx){
    return;
  }

  options.headers = {};
  options.headers['Content-Type'] = "application/x-www-form-urlencoded";
  options.headers['User-Agent'] = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0";
  options.headers['Cookie'] = getCookie(options);

  options.form['dsh'] = dsh[1];
  options.form['GALX'] = galx[1];
  options.form['pstMsg'] = '1';
  options.form['dnConn'] = '';
  options.form['timeStmp'] ='';
  options.form['secTok'] = '';
  options.form['Email'] = _email;
  options.form['Passwd'] = _passwd;
  options.form['signIn'] = '%E3%83%AD%E3%82%B0%E3%82%A4%E3%83%B3';
  options.form['PersistentCookie'] = 'yes';
  options.form['continue'] = 'https%3A%2F%2Fplus.google.com%2F%3Fhl%3Dja%26gpsrc%3Dgplp0';
  options.form['rmShown'] = '1';
  options.form['service'] = 'oz';

  _post(options,onLogin);
}

function onPost(e,r,b) {
  postCb();
}

function onNotification(e,r,b) {

  b = b.replace(new RegExp(/^\)]}\'/),'');
  var j = eval( '(' + b + ')' );
  var notifyItem = j[0][1][1][0];
  var k = [];
  for (var i in notifyItem)
  {
    k[i] = {};
    k[i].id = notifyItem[i][10];
    k[i].date = notifyItem[i][3];
    k[i].unread = notifyItem[i][2][0][1][0][4];
    k[i].comments = [];
    if(notifyItem[i][18][0])
    {
      k[i].name = notifyItem[i][18][0][0][3];
      for(var z in notifyItem[i][18][0][0][7])
      {
        k[i].comments[z] = {};
        k[i].comments[z].name = notifyItem[i][18][0][0][7][z][1];
        k[i].comments[z].comment = notifyItem[i][18][0][0][7][z][2];
        k[i].comments[z].date = notifyItem[i][18][0][0][7][z][3];
        k[i].comments[z].photo = notifyItem[i][18][0][0][7][z][16];
      }
    }
  }
  notifyCb(k);
}

function onNotifyDetail(e,r,b) {
  b = b.replace(new RegExp(/^\)]}\'/),'');
  var j = eval( '(' + b + ')' );

  var activityItem = j[0][1][1][7];
  var k = [];
  for (var i = 0;i<activityItem.length;i++)
  {
    k[i] = {};
    k[i].name = activityItem[i][1];
    k[i].id = activityItem[i][6];
    k[i].date = activityItem[i][3];
    k[i].comment = activityItem[i][2];
  }
  activityCb(k);
}

gPlus.login = function(email,passwd,cb) {
  options = parse(urls['optPreLogin']);
  _email = email;
  _passwd = passwd;
  _loginCb = cb;
  _get(options,onPreLogin);
};

gPlus.post = function(post,cb){

  var scopetype;
  var groupType;
  var me;
  var linkdata;
  var path;
  var opt = {};

  opt = parse(urls['optPost']);
  postCb = cb;

  if(typeof post != 'object')return;
  if(typeof post.circleid == 'undefined'){
    post.circlename = 'anyone';
    post.circleid = '';
    scopetype = 'anyone';
    me  = 'true';
    groupType = null;
  }
  if(typeof post.circlename == 'undefined'){post.circlename = 'anyone';}

  post.message = post.message.replace(new RegExp("\n","g"),"\\n");

  linkdata = "\"[]\"";
  var scopedata = "{\"aclEntries\": [{\"scope\": {\"scopeType\": \"" + scopetype + "\",\"name\": \"" + post.circlename + "\",\"id\": \"" + post.circleid + "\",\"me\":"  + me + ",\"requiresKey\": false, \"groupType\":" + groupType + "},\"role\":20},{\"scope\": {\"scopeType\": \"" + scopetype + "\",\"name\": \"" + post.circlename + "\",\"id\": \"" + post.circleid + "\",\"me\": " + me + ",\"requiresKey\": false, \"groupType\": " + groupType + "},\"role\":60}]}";
  scopedata = scopedata.replace(new RegExp("\"","g"),"\\\"");
  var spar = "[\"" + encodeURI(post.message) + "\",\"oz:" + userid + "." + (parseInt((new Date())/1000,10)).toString(16) + ".0\",null,null,null,null," + linkdata + ",null,\"" + scopedata + "\",true,[],false,false,null,[],false,false,null,null,null,null,null,null,null,null,null,null,null,false,false]";
  opt.headers = {};
  opt.headers['Content-Type'] = "application/x-www-form-urlencoded";
  opt.headers['User-Agent'] = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0";
  opt.headers['Cookie'] = getCookie(options);

  opt.form = {};
  opt.form['spar'] = spar;
  opt.form['at'] = sendid;

  if(typeof post.pageid == 'undefined'){
    opt.path = opt.path + "sharebox/post/?";
  }else{
    opt.path = "/b/" + post.pageid + "/_/sharebox/post/?";
  }
  opt.path = opt.path + "spam=20";
  opt.path = opt.path + "&_reqid=" + randomID(7);
  opt.path = opt.path + "&rt=j";

  _post(opt,onPost);

};

gPlus.getNotification = function(pageid,cb) {
  var opt = {};
  var _pageid ;

  if(typeof pageid == 'function'){
    notifyCb = pageid;
    _pageid = "";
  }
  else
  {
    _pageid = pageid;
    notifyCb = cb;
  }
  opt = parse(urls['optPost']);
  var f_req = "[null,[],5,null,[],null,true,[],null,null,null,null,2,null,null,null,[9]]";
  opt.headers = {};
  opt.headers['Content-Type'] = "application/x-www-form-urlencoded";
  opt.headers['User-Agent'] = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0";
  opt.headers['Cookie'] = getCookie(options);

  if(_pageid === ''){
    opt.path = opt.path + "notifications/getnotificationsdata?";
  }else{
    opt.path = "/b/" + _pageid + "/_/notifications/getnotificationsdata?";
  }
  opt.form = {};
  opt.form['f.req'] = f_req;
  opt.form['at'] = sendid;
  opt.path = opt.path + "hl=ja";
  opt.path = opt.path + "&ozv=es_oz_20130523.13_p5";
  opt.path = opt.path + "&avw=str%3A1";
  opt.path = opt.path + "&fsid=" + fsid;
  opt.path = opt.path + "&_reqid=" + randomID(7);
  opt.path = opt.path + "&rt=j";

  _post(opt,onNotification);
};

gPlus.getNotifyDetail = function(notifyID,pageid,cb) {
  var path;
  var opt = {};
  var _pageid ;

  if(typeof pageid == 'function'){
    activityCb = pageid;
    _pageid = "";
  }
  else
  {
    _pageid = pageid;
    activityCb = cb;
  }
  opt = parse(urls['optPost']);
  var f_req = "[\"" + notifyID + "\",null,null,null,null,null,null,[5,null,null,null,null,null,null,true,null,null,null,null,null,[]]]";
  opt.headers = {};
  opt.headers['Content-Type'] = "application/x-www-form-urlencoded";
  opt.headers['User-Agent'] = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0";
  opt.headers['Cookie'] = getCookie(options);

  if(_pageid === ''){
    opt.path = opt.path + "stream/getactivity/?";
  }else{
    opt.path = "/b/" + _pageid + "/_/stream/getactivity/?";
  }
  opt.form = {};
  opt.form['f.req'] = f_req;
  opt.form['at'] = sendid;

  opt.path = opt.path + "hl=ja";
  opt.path = opt.path + "&ozv=es_oz_20130530.10_p5";
  opt.path = opt.path + "&avw=str%3A1";
  opt.path = opt.path + "&f.sid=" + fsid;
  opt.path = opt.path + "&_reqid=" + randomID(7);
  opt.path = opt.path + "&rt=j";

  _post(opt,onNotifyDetail);
};

_get = function(option,cb){
  option.method = 'GET';
  _request(option,cb);
};

_post = function(option,cb){
  option.method = 'POST';
  _request(option,cb);
};

_request = function (opt,cb)
{

  var form;
  var contents = '';

  if(typeof opt.headers == 'undefined')
    opt.headers = {};
  opt.headers['Content-Type'] = "application/x-www-form-urlencoded";
  opt.headers['User-Agent'] = "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0";
  opt.headers['Cookie'] = getCookie(opt);

  form = opt.form;
  for (var i in form)
  {
    if(contents !== "")
      contents = contents + "&";
    contents = contents + i + '=' + form[i];
  }
  var buffer;
  buffer = '';

  var req = https.request(opt, function(res) {
    cookieStore(res.headers,opt.host);

    res.setEncoding('utf8');
    res.on('data', function (chunk){
      buffer = buffer + chunk;
    });
    res.on('end', function(){
      cb(null,res,buffer);
    });
  });
  req.on('error', function(e) {
    console.error(e);
    cb(e,null,buffer);
  });
  if(opt.method == 'POST')
  {
    req.write(contents);
  }
  req.end();
};
